### ISSUE TITLE
Vsualized region of figure or table by pdfbox

### ISSUE THREAD

**Issue by Sunnycheey** [kermitt2/grobid#528]
Is there some demo code for marking the content out by the coordinate (i.e, p, x, y, w, h) with pdfbox, since I really don't want to go deep with pdf....

Any help will be appreciated.

**Comment by kermitt2** [comment]
Hello @Sunnycheey 

Yes you can have a look at the class `FigureTableVisualizer.java` under `grobid/grobid-core/src/main/java/org/grobid/core/visualization/`. The other classes in the package are good examples too of usage of pdfbox 2.* for extra-annotations with the Grobid results.

**Comment by ehapmgs** [comment]
Any luck in doing so?

I believe it should look something like this based on the code sample


    // grobid coords
    const [page, x, y, h, w] = [6, 177.28, 358.04, 305.92, 8.42]
    // page media box 
    const mediaBox = { lowerX: 0, lowerY: 0, width: 667.28, height: 913.89 }
    const annX = x + mediaBox.lowerX
    const annY = mediaBox.height - (y + h) + mediaBox.lowerY
    const annRightX = x + w + mediaBox.lowerY
    const annTopY = mediaBox.height - y + mediaBox.lowerY
    console.log(annX, annY, annRightX - annX, annTopY - annY)

but it totally points to something else in my case

**Comment by kermitt2** [comment]
@ehapmgs hello! 

Do you want to annotate a PDF with pdfbox (so producing a new PDF with annotations) or annotate a PDF with PDF.js in a browser?

If first, you have examples under `grobid/grobid-core/src/main/java/org/grobid/core/visualization/` as indicated above. 

If second, you are in the wrong issue, but look at the web demo:
https://github.com/kermitt2/grobid/blob/master/grobid-service/src/main/resources/web/grobid/grobid.js#L332 (PDF display)
https://github.com/kermitt2/grobid/blob/master/grobid-service/src/main/resources/web/grobid/grobid.js#L646 (annotation layer on top of the PDF)

Basically you need to scale the annotations (which has no unit) according to the canvas where the PDF is displayed. 

```
var page = thePos.p;
		var pageDiv = $('#page-'+page);
		var canvas = pageDiv.children('canvas').eq(0);;

		var canvasHeight = canvas.height();
		var canvasWidth = canvas.width();
		var scale_x = canvasHeight / page_height;
		var scale_y = canvasWidth / page_width;

		var x = thePos.x * scale_x;
		var y = thePos.y * scale_y;
		var width = thePos.w * scale_x;
		var height = thePos.h * scale_y;
```

**Comment by ehapmgs** [comment]
Hey @kermitt2

Thanks for the help yes it is the first case but it seems there is something wrong with the coords probably a bug that I need to open a new issue for.

 Using the example in  `grobid/grobid-core/src/main/java/org/grobid/core/visualization/` you can see the annotations are not located correctly 

![Screen Shot 2020-12-03 at 6 59 51 PM](https://user-images.githubusercontent.com/26691494/101061958-bce00b00-3599-11eb-8231-154423e4649c.png)

I believe it used to work in version `0.5.4` I can see that the coords for the same figure is different between the `master` branch  and `0.5.4`

**Comment by kermitt2** [comment]
Indeed thanks ! I didn't look at this part since very long, I am re-discovering it :) 

There are actually web service to get the annotated PDF:

```console
curl -v --form input=@./s41523-020-00198-1.pdf --form type=2 localhost:8070/api/annotatePDF > s41523-020-00198-1-annot.pdf
```

s41523-020-00198-1-annot.pdf gives the PDF with annotations, the attribute `type` indicates which annotations to put (`0`: citations, `1`: blocks, `2`: figure).

This is working on some PDF:

![Screenshot from 2020-12-03 18-46-24](https://user-images.githubusercontent.com/2340795/101067463-dd0eca80-3597-11eb-90f8-4729a931a13c.png)

But for the one you tested, there is a scaling problem. I think Achraf fixed this problem in pdfalto with https://github.com/kermitt2/pdfalto/issues/43 - see https://github.com/kermitt2/grobid/issues/330 

So just need to do the same with pdfbox here... mmm

**Comment by kermitt2** [comment]
For reference, failing PDF https://onlinelibrary.wiley.com/doi/pdf/10.1087/20100308

**Comment by kermitt2** [comment]
Normally fixed with 33b50b7fe5aad8620be681c2c910ffdbb9a27573

![Screenshot from 2020-12-03 19-07-59](https://user-images.githubusercontent.com/2340795/101070250-62e04500-359b-11eb-8909-ac63c475017e.png)

**Comment by ehapmgs** [comment]
@kermitt2 nice! it is working as expected now thanks
